---
title: "PRS predictions"
author: "Arslan Zaidi"
date: "6/25/2020"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

We are going to simulate a frequency-dependent genetic architecture for the GWAS. To do that, I'm going to try out the model proposed in Schoech et al. 2019. 

This is specifically for the out of sample test (pheno file 3).

## Model

The effect size of a variant i, 

$$\beta_i \sim N( 0, \sigma_{l}^2 . [p_i.(1-p_i)]^\alpha )$$

where:

$\sigma_l$ = the frequency-independent component of the genetic variance associated with variant $i$, 

$p_i$ = the frequency of the $i_{th}$ allele,

$\alpha$ = the scaling factor which determines how the effect size is related to allele frequency. 

Under this model, the total contribution of variant $i$ to the genetic variance is

$$\sigma_{i}^2 = \sigma_l^2.[2.p_i.(1-p_i)]^{\alpha+1}$$

## Explanation

$$\sigma_{g}^2 = Var(\beta_iX_i) = \mathbb{E}[\beta_i^2X_i^2] - \mathbb{E}[\beta_iX_i]^2$$
$$ = \mathbb{E}[\beta_i^2].\mathbb{E}[X_i^2] - (\mathbb{E}[\beta_i].\mathbb{E}[X_i])^2$$
$$ = \sigma_l^2.2p_i(1-p_i)[p_i(1-p_i)]^\alpha$$
$$ = \sigma_l^2.[2p_i(1-p_i)]^{1+\alpha}$$

And the total additive genetic variance across $M$ variants is

$$\sigma_{g}^2 = \sum_{i=1}^M\sigma_{i}^2= \sum_{i=1}^M\sigma^2_{l}.[2.p_i.(1-p_i)]^{\alpha+1}$$

We would like to choose effect sizes such that the total heritability is 0.8. To do this, we can set the phenotypic variance to 1, and $\sigma_g^2$ to 0.8.

## Implementation

```{r libraries, message=FALSE}

library(ggplot2)
library(data.table)
library(dplyr)
library(rprojroot)

F=is_rstudio_project$make_fix_file()

```

I thinned down variants from the msprime simulations such that there was only 1 variant per 100kb. This ensured that the (estimated) effect sizes are not compounded across loci due to LD. This yielded 2000 variants. I am going to emulate the architecture for height. Shoech et al. 2019 estimate $\alpha = -0.4$ so that's what I'm going to use in my simulations.I'm also going to restrict the heritability to be around 0.8 (similar to what's known for height).

To do this, I must estimate $\sigma_l$, the frequency-independent component of variance. We know that: $\sigma_{g} = \sum_{i=1}^M \beta^2_i.[2p(1-p)]$. To calculae $\sigma^2_l$, that will give us $\mathbb{E}[\sigma_{g}^2] = 0.8$, we replace $\beta_i^2$ with $\mathbb{E}[\beta^2_i]$. Thus, 

$$\mathbb{E}[\beta^2_i] = \sigma^2_l[2p_i(1-p_i)^{\alpha}]$$
$$\sigma_g^2 = \sum_{i=1}^M \sigma_l^2[2p_i(1-p_i)]^{1+\alpha} = 0.8$$


$$ \sigma_{g}^2 = \sigma^2_l\sum_{i=1}^M[2p_i(1-p_i)]^{\alpha+1} = 0.8$$
$$\sigma^2_l = \frac{0.8}{\sum_{i=1}^M[2p_i(1-p_i)]^{\alpha+1}}$$

```{r sim_effects}

set.seed(123)

# load variant frequency file
p<-fread(F("gwas/grid/genotypes/tau100/ss500/train/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.train.snps.thinned_100kb.afreq"))


#calculate the independent component of variance required
sigma2_l=0.8/sum(sapply(p$ALT_FREQS,function(x){
  beta=(2*x*(1-x))^(1-0.4)
  return(beta)
}))

#sample maf-dependent effects using the model above
p$beta=sapply(p$ALT_FREQS,function(x){
  beta=rnorm( 1 , mean=0, sd=sqrt(sigma2_l * (2*x*(1-x))^-0.4 ))
})

#let's calculate sigma2_g to confirm that the total genetic variance is indeed 0.8
sigma2_g=sum(mapply(function(b,p){ b^2* 2*p*(1-p) }, p$beta, p$ALT_FREQS))
paste("sigma2_g:", sigma2_g)


#let's plot the effect sizes to confirm and calculate the overall genetic variance

ggplot(p,aes(ALT_FREQS,beta))+
  geom_point(size=0.6)+
  theme_bw()+
  labs(x=bquote(p[i]),
       y=bquote(beta[i]))+
  scale_x_log10()+
  theme(panel.grid = element_blank())

```

The genotypic value for all individuals can be generated as $X^T \beta$, where $X$ is an $n \times m$ matrix of genotypes and $\beta$ is a $m \times 1 $ column vector of effect sizes. Write the effect sizes to file to generate the genetic values using PLINK.

```{r write_effects, eval=FALSE}

#save the effect sizes to file and use plink2 to generate PRS

fwrite(p%>%
         select(ID,ALT,beta),
           F("gwas/grid/genotypes/tau100/ss500/train/genos_grid_d36_m0.05_s500_t100.rmdup.train.thinned_100kb.effects"),
       row.names=FALSE,col.names=FALSE,quote=FALSE,sep="\t")


```

Read the PRS generated by plink2 and add residual (random) variation to acquire desired heritability.

```{r read_prs}

prs=fread(F(
  "gwas/grid/genotypes/tau100/ss500/train/genos_grid_d36_m0.05_s500_t100.rmdup.train.thinned_100kb.gvalue.sscore"
))

colnames(prs)<-c("iid","dosage","prs")


```


Now add some 'noise' to the data to get desired heritability. Because the genetic variance is 0.8, I'm going to add random effects as: $\epsilon \sim N(0,0.2)$.

For phenotypes with an environmental (sharp or smooth) contribution, the model will be:

$$Y = X^T\beta + \mu  + \epsilon$$
where $\mu$ is the environmental effect, which depends on the deme an individual belongs to.

I will generate three phenotypes:
1. random: No environmental effect.
2. smooth: smooth cline from north to south where the difference in the average effect of the northernmost demes and southermost demes is 2 sd.
3. sharp: the 3rd deme is picked (at random) and given an effect of 2 sd whereas the other demes have no effect.


```{r sim_phenotypes}

#load file containing the deme id and latitude and longitude for each individual
#this will be 
pheno=fread(F("gwas/grid/genotypes/tau100/ss500/iid_train.txt"))

prs=merge(prs,pheno[,c('IID','deme','latitude','longitude')],by.x="iid",by.y="IID",sort=FALSE)


#no 'environmental' effect
prs$grandom = rnorm(9000,0, sqrt(1-sigma2_g))

#smooth effect
prs$smooth = sapply(prs$latitude,
                      function(x){
                        rnorm(n=1,
                              mean=(x+1)/3,
                              sd=sqrt(1-sigma2_g))})

#sharp environmental effect
prs$sharp = sapply(prs$deme,
                 function(x){
                   if(x==2){
                     rnorm(n=1,
                           mean=2,
                           sd=sqrt(1-sigma2_g))}else{
                             rnorm(n=1,
                                   mean=0,
                                   sd=sqrt(1-sigma2_g))
                           }})

#scale each so that their variances are 0.2 - to adjust heritability to 0.8
prs$grandom=scale(prs$grandom,scale=TRUE)*sqrt(1-sigma2_g)
prs$sharp=scale(prs$sharp,scale=TRUE)*sqrt(1-sigma2_g)
prs$smooth=scale(prs$smooth,scale=TRUE)*sqrt(1-sigma2_g)

#add prs to each of the environmental effects
prs=prs%>%
  mutate(grandom=prs+grandom,
         smooth=prs+smooth,
         sharp=prs+sharp,
         random=rnorm(9000,0,1))

#plot the spatial distribution of the phenotypes
mprs=prs%>%
  select(iid,deme,latitude,longitude,grandom,random,smooth,sharp)%>%
  group_by(deme,latitude,longitude)%>%
  summarize(grandom=mean(grandom),
            random=mean(random),
            smooth=mean(smooth),
            sharp=mean(sharp))%>%
  melt(id.vars=c("deme","longitude","latitude"))


ggplot(mprs,aes(longitude,latitude,fill=value))+
  geom_tile()+
  facet_wrap(~variable,scales="free")+
  scale_fill_viridis_c()+
  theme_bw()

```


```{r write_phenotypes, eval=F}

fwrite(
  prs%>%
    mutate(FID=iid,IID=iid)%>%
  select(FID,IID,prs,grandom,random,smooth,sharp),
  F("gwas/grid/test_genos_grid_d36_m0.05_s500_t100_gl200_thinned_100kb.wtge.pheno3"),
  col.names=TRUE,row.names=FALSE,quote=FALSE,sep="\t")
```

