---
title: "GWAS in a spatially structured population - grid"
author: "AAZaidi"
date: "11/7/2019"
output: html_document
---


```{r}
library(data.table)
library(ggplot2)
library(here)
library(cowplot)
library(dplyr)
library(tidyr)

```


## Define important functions

```{r}


#function that:
#1. reads in GWAS association (tau=100)
#2. calculates expected -log10pvalues
#3. subsamples for plotting
read_gwas<-function(phenotype,pcs=0,freq="common"){
  print("reading association results")
  if(pcs==0){
  df=fread(here(
    paste(
      "gwas/grid/gwas_results/tau-9/e2/gwas_grid_d36_m0.07_s250_t-9_gl200_e2_pcs0.all.",phenotype,".glm.linear",
      sep="")
    ))
  
  #df[,c("phenotype","pcfreq","pcs") := list(phenotype,"No correction",paste("nPCS:",0))]

  }else{
    df=fread(here(paste(
      "gwas/grid/gwas_results/tau-9/e2/gwas_grid_d36_m0.07_s250_t-9_gl200_e2_",freq,".pcs",pcs,".all.",phenotype,".glm.linear" 
      ,sep="")
    ))

  #df[,c("phenotype","pcfreq","pcs") := list(phenotype,"No correction",paste("nPCS:",pcs))]
  
  }
  
  colnames(df)[1]="CHROM"
  df=df[,c("ID","P")]
  
  print("adding frequency information for variants")
  df<-cbind(df,fcat_df)
  colnames(df)[ncol(df)]="fcat"
  
  df=df[fcat%in%c("common","rare"),]
  
  #df[,fcat:=cut(ALT_FREQS, breaks=c(0,0.05,1), labels=c("rare","common"))]
  
  print("calculating expected Pvalue distribution and lambda")
  
  df=df[order(P),.SD,by="fcat"]
  df[,c("exp.p","obs.chi"):=list(ppoints(length(ID)), qchisq(P,df=1,lower.tail = F)), by="fcat"]
  df[,"exp.chi":=qchisq(exp.p,df=1,lower.tail = F)]
  df[,"lambda":=obs.chi/exp.chi]
  df=df[,chi.percentile:=(length(exp.p):1)/length(exp.p),by="fcat"]
  
    
  # df<-as.data.table(df%>%
  #   mutate(lower.ci=qbeta(0.025,shape1=row_number(),shape2 = length(ID)-row_number()),
  #          upper.ci=qbeta(0.975,shape1=row_number(),shape2 = length(ID)-row_number()))%>%
  #     ungroup())
  
  print("reducing the number of rows for plotting")
  #keep the first 1000 SNPs (low-p tails) and subsample from the rest (only for rare variants, keep all common)
  df.common=df[fcat=="common",]
  df=df[fcat=="rare",]
  df=rbind(df.common,
           df[c(1:1000),],
           df[seq(1001,nrow(df),1000)]
           )
  
  return(df)
}

##function for plotting
fplt=function(df,max.lambda){
  
  plt1<-ggplot(data=df)+
  geom_ribbon(data=mlambda,aes(x=-log10(exp.p),
                               ymin=-log10(lower.ci),
                               ymax=-log10(upper.ci),
                               fill=fcat),
              alpha=0.2,
              show.legend = F)+
  geom_line(aes(-log10(exp.p),
                 -log10(P),
                 color=fcat),
             size=1,
             alpha=0.5,show.legend = F)+
  geom_abline(intercept=0,slope=1,color="black")+
  scale_color_manual(values=c("#ff7f00","#377eb8"))+
  scale_fill_manual(values=c("#ff7f00","#377eb8"))+
  theme_bw()+
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14),
        axis.title=element_blank())+
  labs(x=expression(-log[10]~Expected~pvalue),
       y=expression(-log[10]~Observed~pvalue),
       color="Variant")+
  xlim(c(0,9))+
  ylim(c(0,9))
  
  plt.inset=ggplot()+
    geom_line(data=df,aes(chi.percentile,
                lambda,
                color=fcat),
            show.legend = F,
            size=0.5)+
    annotate(geom="text",
             x=0.9991,
             y=0.9*max.lambda,
             label="lambda[p]",parse=T)+
  theme_bw()+
  theme(panel.grid.major.x = element_blank(),
        legend.position="none",
        axis.title.y=element_blank(),
        panel.grid=element_blank(),
        plot.background = element_blank(),
        axis.text.x=element_text(hjust=0))+
  scale_x_log10(limits=c(0.999,1),
                breaks=c(0.999,1),
                labels=c("0.999","1"))+
  scale_y_continuous(limits=c(0.99,max.lambda),
                     breaks=c(1,max.lambda),
                     position="right")+
      labs(x="p")+
    scale_color_manual(values=c("#ff7f00","#377eb8"))
  
  plt.wt.inset<-ggdraw(plt1) +
  draw_plot(plt.inset, .17, .60, .5, .35)
  
  return(plt.wt.inset)

}


```


## Read allele frequencies for tau=-9. 

```{r}

freq9<-fread(here("gwas/grid/genotypes/tau-9/genos_grid_d36_m0.07_s250_t-9_gl200.frq.afreq"))

freq9[,fcat:=cut(ALT_FREQS, breaks=c(0,0.01,0.05,1), labels=c("rare","medium","common"))]
fcat_df=freq9$fcat

snp.counts=freq9[,.(counts=length(ALT_FREQS)),by="fcat"]
snp.cm=as.numeric(snp.counts[2,2])
snp.re=as.numeric(snp.counts[1,2])

rm(freq9)

rlunif <- function(n, min, max, base=exp(1)) {
  if (mode(n) != "numeric")
    stop("'n' must be a non-empty numeric vector")
  if (any(missing(min), missing(max)))
    stop("'min' and 'max' not provided, without default.\n")
  ifelse(base == exp(1),
         return(exp(runif(n, log(min, base), log(max, base)))),
         return(base ^ (runif(n, log(min, base), log(max, base)))))
}

#generate CI for -log10(P) vaues
lambda1=data.frame(common=c(1:1000,round(log(rlunif(4000,min=exp(1e-08),max=exp(0.99)))*snp.cm)),
                  rare=c(1:1000,round(log(rlunif(4000,min=exp(1e-08),max=exp(0.99)))*snp.re)))

mlambda=melt(lambda1)
colnames(mlambda)=c("fcat","ix")

mlambda=mlambda%>%
  group_by(fcat)%>%
  arrange(ix)%>%
  mutate(exp.p=ix/max(ix),
         lower.ci=qbeta(0.025,shape1=ix,shape2 = max(ix)-ix),
         upper.ci=qbeta(0.975,shape1=ix,shape2 = max(ix)-ix))

mlambda$fcat=factor(mlambda$fcat,levels=c("rare","common"))


```

## Plot GWAS QQplots

### Smooth environment and no correction.

```{r}

sm_fcommon_pc0_t9=read_gwas(phenotype = "smooth",pcs = 0,freq = "common")

plt_sm_pc0_t9=fplt(sm_fcommon_pc0_t9,max.lambda=1.19)

rm(sm_fcommon_pc0_t9)
```

### Smooth environment and correction for 100 PCs (common variants - MAF>0.01)

```{r}

sm_fcommon_pc100_t9=read_gwas(phenotype = "smooth",pcs = 100,freq = "common")

plt_sm_fcommon_pc100_t9<-fplt(sm_fcommon_pc100_t9, max.lambda=1.19)
  
rm(sm_fcommon_pc100_t9)


```

### Smooth effect and correction with rare PCA (100 PCs)

```{r}

sm_frare_pc100_t9=read_gwas(phenotype = "smooth",pcs = 100,freq = "rare")

plt_sm_frare_pc100_t9<-fplt(sm_frare_pc100_t9, 1.19)


rm(sm_frare_pc100_t9)

```

### Sharp environmental effect and no correction

```{r}

shp_fcommon_pc0_t9=read_gwas(phenotype = "sharp",pcs = 0)

plt_shp_pc0_t9=fplt(shp_fcommon_pc0_t9,1.34)

rm(shp_fcommon_pc0_t9)

```

### Sharp environment and correction for 100 PCs (common variants - MAF>0.01)

```{r}

shp_fcommon_pc100_t9=read_gwas(phenotype = "sharp",pcs = 100,freq = "common")

plt_shp_fcommon_pc100_t9=fplt(shp_fcommon_pc100_t9,1.34)


rm(shp_fcommon_pc100_t9)


```

#### Sharp effect and correction with rare PCA (100 PCs)

```{r}

shp_frare_pc100_t9=read_gwas(phenotype = "sharp",pcs = 100,freq = "rare")

plt_shp_frare_pc100_t9=fplt(shp_frare_pc100_t9,1.34)

rm(shp_frare_pc100_t9)


```




```{r}

plts.arranged<-plot_grid(plt_sm_pc0_t9,
                         plt_sm_fcommon_pc100_t9,
                         plt_sm_frare_pc100_t9,
                         plt_shp_pc0_t9,
                         plt_shp_fcommon_pc100_t9,
                         plt_shp_frare_pc100_t9,
          ncol=3,
          nrow=2,
          rel_widths = rep(1,3)/3)


plts.arranged

ggsave(here("gwas/grid/analysis/Single_association_statistic/plts_singlesnpgwas_combined_t9.pdf"),
       height=7,width=7)
 
```








