---
title: "R Notebook"
output: html_notebook
---

```{r}

library(ggplot2)
library(data.table)
library(tidyr)
library(dplyr)
library(rprojroot)

F = is_rstudio_project$make_fix_file()

starts = seq(0,99)*1e5 + 1
stops = starts + 1e5

isin = function(x){
  for(i in 1:length(starts)){
    if(x>starts[i] & x<stops[i]){
      return(i)
    }
  }
}

add_window = function(dd){
  
  dd = dd%>%
    separate(SNP,into=c("chrom","POS","REF","alt"),
             remove = FALSE,
             sep="_")
  dd$POS = as.numeric(dd$POS)
  colix =which(!(colnames(dd) %in% c("REF","alt")))
  dd = dd[ , colix , with = FALSE]
  dd$window = sapply(dd$POS, isin)
  dd$window = paste(dd$chrom,"_",dd$window,sep="")
  return(dd)
}


true.effects = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/true_effects/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.train.all.thinned_100kb.effects"))

colnames(true.effects) = c("rep","SNP","A1","esize")
true.effects = add_window(true.effects)

```

Let's first check the approach where we ascertain in standard GWAS (train), re-estimate in another set of unrelated individuals (one used to generate sibling haplotypes), then predict in a third (test). 

```{r}

betas1 = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/betas/effects.smooth.a1_r2.pcs0.all.betas"))

colnames(betas1) =c("rep","SNP","A1","pcs0","cm","re","cmre")

#let's check the correlation between effect size and true effects
#to do this we need to assign each variant to a window
betas1 = add_window(betas1)
betas1 = merge(betas1, true.effects%>%
                 select(rep,chrom,window,SNP,esize),
               by = c("rep","window"))

ggplot(betas1,aes(esize,pcs0))+
  geom_point()

#calculate correlation between true and estimated effect size
betas1%>%
  group_by(rep)%>%
  summarize(rcor = cor(abs(esize),abs(pcs0)))%>%
  pull(rcor)%>%mean()


#the prs was then re-estimated in the third sample
#load the frequency of this sample
freq3 = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/freqs/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.test.snps.frq.afreq.gz"))

colnames(freq3)=c("chrom","SNP","ref","alt","frq_test","ct")
freq3 = freq3[,.(SNP,frq_test)]

#let's calculate the expected genetic values in the test sample
g.test = merge(true.effects,freq3,by="SNP")
g.test = g.test%>%
  mutate(g.exp = 2*esize*frq_test)

betas1 = merge(betas1,freq3,by.x="SNP.x",by.y="SNP")
betas1 = betas1%>%
  mutate(g.hat = 2*pcs0*frq_test)

betas1 = merge(betas1,
               g.test%>%
                 select(rep,SNP,g.exp,window),by=c("rep","window"))

betas1 = betas1%>%
  mutate(maf_test = case_when(frq_test <= 0.5 ~ frq_test,
                                 TRUE ~ 1 - frq_test))

gvalues.test = betas1%>%
  group_by(rep)%>%
  summarize(g.hat = sum(g.hat),
            g.exp = sum(g.exp))

ggplot(betas1%>%filter(pcs0!=0),aes(g.exp,g.hat,color=maf_test))+
  geom_point()


```

