---
title: "R Notebook"
output: html_notebook
---

```{r}

library(ggplot2)
library(data.table)
library(tidyr)
library(dplyr)
library(rprojroot)

F = is_rstudio_project$make_fix_file()

```


```{r}

true.effects = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/true_effects/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.train.all.thinned_100kb.effects"))

colnames(true.effects) = c("rep","SNP","A1","esize")
true.effects = true.effects%>%
  separate(SNP,into=c("chrom","POS","REF","ALT"),
           remove = FALSE,
           split="_")

true.effects$POS = as.numeric(true.effects$POS)

true.effects = true.effects %>%
  select(rep,SNP,POS,A1,esize)
  
train.freq = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/freqs/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.train.snps.frq.afreq.gz"))

colnames(train.freq) = c("CHROM","SNP","REF","ALT","ALT_FREQS","CT")

train.freq = train.freq[,.(CHROM,SNP,ALT,ALT_FREQS)]

true.effects = merge(true.effects, train.freq, by = "SNP")

ggplot(true.effects,
       aes(ALT_FREQS))+
  geom_histogram()+
  facet_wrap(~rep)+
  scale_x_log10()+
  theme_classic()

```

Variants ascertained in standard GWAS and re-estimated in an independent GWAS. 

```{r}

betas1 = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/betas/effects.smooth.a1_r2.pcs0.all.betas"))

colnames(betas1) =c("rep","SNP","A1","pcs0","cm","re","cmre")

#the prs was then re-estimated in the third sample

betas1 = merge(betas1,train.freq, by ="SNP")

#calculate number of variants used in prs calculation
betas1%>%
  filter(pcs0!=0)%>%
  group_by(rep)%>%
  summarize(nsnps = n())%>%
  pull(nsnps)%>%mean()

ggplot(betas1%>%
         filter(pcs0 !=0),
       aes(ALT_FREQS))+
  geom_histogram()+
  facet_wrap(~rep)+
  scale_x_log10()+
  theme_classic()

```

```{r}

betas2 = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/betas/effects.smooth.a3s_r1.pcs0.all.betas"))

colnames(betas2) =c("rep","SNP","A1","causal","causal_p","lead")

#calculate number of variants used in prs calculation
betas2%>%
  filter(lead!=0)%>%
  group_by(rep)%>%
  summarize(nsnps = n())%>%
  pull(nsnps)%>%mean()


betas2 = merge(betas2,true.freq, by ="SNP")

ggplot(betas2%>%
         filter(lead !=0),
       aes(ALT_FREQS))+
  geom_histogram()+
  facet_wrap(~rep)+
  scale_x_log10()+
  theme_classic()

```



```{r}

ggplot()+
  geom_histogram(data = betas2%>%
                   filter(lead != 0),
                 aes(ALT_FREQS),
                 fill = "orange", alpha=0.4)+
  geom_histogram(data = betas1%>%
                   filter(pcs0 != 0),
                 aes(ALT_FREQS),
                 fill="blue", alpha=0.4)+
  facet_wrap(~rep)+
  scale_x_log10()+
  theme_classic()

```


```{r}

betas3 = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/betas/effects.smooth.a3s_r1.re.all.betas"))

colnames(betas3) =c("rep","SNP","A1","causal","causal_p","lead")

#calculate number of variants used in prs calculation
betas1%>%
  filter(pcs0!=0)%>%
  group_by(rep)%>%
  summarize(nsnps = n())%>%
  pull(nsnps)%>%mean()

betas3 = merge(betas3,true.freq, by ="SNP")


```


```{r}

betas4 = fread(F("gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/betas/effects.smooth.a1_r3s.ncorr.all.betas"))

colnames(betas4) =c("rep","SNP","A1","pcs0","cm","re","cmre")

#calculate number of variants used in prs calculation
betas4%>%
  filter(pcs0!=0)%>%
  group_by(rep)%>%
  summarize(nsnps = n())%>%
  pull(nsnps)%>%mean()

betas4 = merge(betas4,true.freq, by ="SNP")

betas4 = betas4%>%
  separate(SNP,into=c("chrom","POS","REF","alt"),
           remove = FALSE,
           sep="_")

betas4$POS = as.numeric(betas4$POS)
betas4$window = sapply(betas4$POS, isin)

betas4 = merge(betas4, true.effects,
               by=c("rep","CHROM","window"))

ggplot(betas4%>%
         filter(pcs0!=0),aes(esize,pcs0))+
  geom_point()+
  stat_smooth(method="lm")

var4 = betas4%>%
  group_by(rep)%>%
  summarize(var.x = sum(pcs0^2 * ALT_FREQS.x*(1-ALT_FREQS.x)),
            var.y = sum(esize^2 * ALT_FREQS.y*(1-ALT_FREQS.y)))

```



```{r}

starts = seq(0,99)*1e5 + 1
stops = starts + 1e5

isin = function(x){
  for(i in 1:length(starts)){
    if(x>starts[i] & x<stops[i]){
      return(i)
    }
  }
}

true.effects$window = sapply(true.effects$POS,
                             isin)


betas1 = betas1%>%
  separate(SNP,into=c("chrom","POS","REF","alt"),
           remove = FALSE,
           sep="_")

betas1$POS = as.numeric(betas1$POS)
betas1 = betas1[,.(CHROM,POS,rep,A1,pcs0,cm,re,cmre,ALT_FREQS)]
betas1$window = sapply(betas1$POS, isin)

betas1 = merge(betas1, true.effects,
               by=c("rep","CHROM","window"))


ggplot(betas1%>%
         filter(pcs0!=0),aes(esize,pcs0))+
  geom_point()+
  stat_smooth(method="lm")+
  facet_wrap(~rep)

betas2 = betas2%>%
  separate(SNP,into=c("chrom","POS","REF","alt"),
           remove = FALSE,
           sep="_")

betas2$POS = as.numeric(betas2$POS)
betas2 = betas2[,.(CHROM,POS,rep,A1,pcs0,cm,re,cmre,ALT_FREQS)]
betas2$window = sapply(betas2$POS, isin)

betas2 = merge(betas2, true.effects,
               by=c("rep","CHROM","window"))

ggplot(betas2%>%
         filter(lead!=0),aes(esize,lead))+
  geom_point()+
  stat_smooth(method="lm")


betas3 = betas3%>%
  separate(SNP,into=c("chrom","POS","REF","alt"),
           remove = FALSE,
           sep="_")

betas3$POS = as.numeric(betas3$POS)
betas3 = betas3[,.(CHROM,POS,rep,A1,pcs0,cm,re,cmre,ALT_FREQS)]
betas3$window = sapply(betas3$POS, isin)

betas3 = merge(betas3, true.effects,
               by=c("rep","CHROM","window"))

ggplot(betas3%>%
         filter(lead!=0),aes(esize,lead))+
  geom_point()+
  stat_smooth(method="lm")



var1 = betas1%>%
  group_by(rep)%>%
  summarize(var.x = sum(pcs0^2 * ALT_FREQS.x*(1-ALT_FREQS.x)),
            var.y = sum(esize^2 * ALT_FREQS.y*(1-ALT_FREQS.y)))
  
var2 = betas2%>%
  group_by(rep)%>%
  summarize(var.x = sum(lead^2 * ALT_FREQS.x*(1-ALT_FREQS.x)),
            var.y = sum(esize^2 * ALT_FREQS.y*(1-ALT_FREQS.y)))

var3 = betas3%>%
  group_by(rep)%>%
  summarize(var.x = sum(lead^2 * ALT_FREQS.x*(1 - ALT_FREQS.x)),
            var.y = sum(esize^2 * ALT_FREQS.y*(1 - ALT_FREQS.y)))




```


