<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Examining prediction accuracy as a consequence of effect sizes estimated in the discovery vs an independent sample</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Demographic history mediates the effect of stratification on polygenic scores</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Arslan-Zaidi/popstructure">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Examining prediction accuracy as a consequence of effect sizes estimated in the discovery vs an independent sample</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-12-22
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>popstruct_scripts/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201202code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201202)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201202code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201202)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomArslanZaidipopstructuretree99813ec398a9cad0bb1775140d09a7ab2cc431fctargetblank99813eca"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/Arslan-Zaidi/popstructure/tree/99813ec398a9cad0bb1775140d09a7ab2cc431fc" target="_blank">99813ec</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomArslanZaidipopstructuretree99813ec398a9cad0bb1775140d09a7ab2cc431fctargetblank99813eca" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/Arslan-Zaidi/popstructure/tree/99813ec398a9cad0bb1775140d09a7ab2cc431fc" target="_blank">99813ec</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rproj.user/
    Ignored:    Archived/
    Ignored:    analysis/.DS_Store
    Ignored:    burden_msprime/.DS_Store
    Ignored:    burden_msprime/generate_burden/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/burden_msprime/
    Ignored:    data/burden_msprime2/
    Ignored:    data/gwas/
    Ignored:    data/ukmap/
    Ignored:    gwas/.DS_Store
    Ignored:    gwas/grid/.DS_Store
    Ignored:    gwas/grid/tau-9/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/generate_genotypes/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/gwas/.DS_Store
    Ignored:    gwas/grid/tau100/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/generate_genotypes/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/gwas/.DS_Store
    Ignored:    gwas/ukb/.DS_Store
    Ignored:    gwas/ukb/scripts/.DS_Store
    Ignored:    optimize_migration_rate/.DS_Store
    Ignored:    optimize_migration_rate/complex_dem/.DS_Store
    Ignored:    optimize_migration_rate/grid/.DS_Store
    Ignored:    output/plots/
    Ignored:    pca_plots/.DS_Store
    Ignored:    prs/.DS_Store
    Ignored:    prs/grid/.DS_Store
    Ignored:    qqplots/.DS_Store
    Ignored:    qqplots/fixed_effects/.DS_Store
    Ignored:    qqplots/lmms/.DS_Store
    Ignored:    revisions/.DS_Store
    Ignored:    shared_scripts/gwas/
    Ignored:    shared_scripts/prs/
    Ignored:    shared_scripts/simphenotype/
    Ignored:    sib_analysis/.DS_Store
    Ignored:    sib_analysis/complex_dem/.DS_Store
    Ignored:    sib_analysis/grid/.DS_Store
    Ignored:    sib_analysis/grid/tau100/mates/
    Ignored:    simulating_genotypes/.DS_Store
    Ignored:    simulating_genotypes/grid/.DS_Store
    Ignored:    simulating_phenotypes/.DS_Store

Untracked files:
    Untracked:  analysis/biasvaccuracy_prsascertainment.Rmd
    Untracked:  analysis/biasvaccuracy_prsascertainment.nb.html
    Untracked:  analysis/plotting_prs_sib_effects.Rmd
    Untracked:  analysis/plotting_prs_sib_effects.nb.html
    Untracked:  analysis/plottingprs_distribution_gridt.Rmd
    Untracked:  analysis/plottingprs_distribution_gridt.nb.html
    Untracked:  analysis/plt_burden_association.Rmd
    Untracked:  analysis/plt_ukb_unrelated_prs.Rmd
    Untracked:  analysis/plt_ukb_unrelated_prs.nb.html
    Untracked:  analysis/prs_wt_finemapping.Rmd
    Untracked:  analysis/prs_wt_finemapping.nb.html

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd
    Modified:   burden_msprime/plt_burden_association_t100.Rmd
    Modified:   burden_msprime/plt_burden_association_t100.nb.html
    Deleted:    revisions/figuring_out_prediction_accuracy.Rmd
    Deleted:    revisions/figuring_out_prediction_accuracy.nb.html
    Deleted:    revisions/figuring_out_prediction_accuracy2.Rmd
    Deleted:    revisions/figuring_out_prediction_accuracy2.nb.html

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>We were seeing that prediction accuracy, measured as the correlation between the polygenic score and genetic value, was much higher (~2x) when the variants were discovered in one sample (N = 9K) but the effects were re-estimated in siblings (N = 9k). This prediction accuracy was even higher than a fully siblig gwas (also 9k) where presumably the effects are more unbiased (not as impacted by stratification).</p>
<p>We think this may have something to do with winner’s curse or the fact that the increase in accuracy is due to the fact that the effects are re-estimated in an independent sample. To test this, let’s ignore the siblings and calculate PRS in two ways:</p>
<ol style="list-style-type: decimal">
<li><p>effects estimated in the discovery sample (this is what is normally done).</p></li>
<li><p>variants discovered in a GWAS in unrelated individuals and effects re-estimated in an independent sample.</p></li>
</ol>
<p>Let’s calculate both bias and prediction accuracy in both ways.</p>
<pre class="r"><code>library(ggplot2)
library(data.table)
library(dplyr)</code></pre>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:data.table&#39;:

    between, first, last</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>library(rprojroot)
library(patchwork)

F = is_rstudio_project$make_fix_file()

options(dplyr.summarise.inform=FALSE) </code></pre>
<div id="effects-estimated-in-the-same-sample-as-the-discovery-set" class="section level3">
<h3>Effects estimated in the same sample as the discovery set</h3>
<p>Question: What is the accuracy of polygenic risk prediction when we estimate effects in the same sample as the discovery set?</p>
<p>Plot the bias in polyegenic score measured by the correlation between residual polygenic score and latitude (the confounding environmental variable).</p>
<pre class="r"><code>#effects discovered and estimated in training set and
#prs predicted in 3rd set (used to construct sibling haplotypes)
prs1 = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/prs1sample/prs.smooth.a1_p3.all.nc.sscore&quot;))

colnames(prs1) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;pcs0&quot;,&quot;cm&quot;,&quot;re&quot;,&quot;cmre&quot;)

mprs1 = reshape2::melt(prs1,id.vars = c(&quot;rep&quot;,&quot;IID&quot;),
                       value.name = &quot;prs&quot;,
                       variable.name = &quot;correction&quot;)

#load genetic values for individuals in the sample we are predicting - to calculate prediction accuracy
gvalue1 = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/gvalues/gvalue.p3.all.sscore&quot;))

colnames(gvalue1) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;gvalue&quot;)

#load latitude information - to calculate bias
pop1 = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/iid_sib.txt&quot;))

#add latitude info
mprs1 = merge(mprs1, pop1, by=&quot;IID&quot;)
#add genetic value
mprs1 = merge(mprs1, gvalue1, by=c(&quot;rep&quot;,&quot;IID&quot;))

#center the prs and subtract out genetic value
mprs1 = mprs1%&gt;%
  group_by(rep,correction)%&gt;%
  mutate(prs.adjusted = prs - mean(prs),
         prs.adjusted = prs.adjusted - gvalue)

#calculate the bias and prediction accuracy 
mprs1.bias = mprs1 %&gt;%
  group_by(rep,correction)%&gt;%
  summarize(rlat = cor(prs,latitude),
            r2 = cor(prs,gvalue)^2)

plt_bias.all = ggplot(mprs1.bias,aes(rlat))+
  geom_histogram(bins=10)+
  facet_wrap( ~ correction,
              labeller = as_labeller(c(
                pcs0 = &quot;No correction&quot;,
                cm = &quot;Common-PCA&quot;,
                re = &quot;Rare-PCA&quot;,
                cmre = &quot;Common + rare&quot;
              )))+
  theme_classic()+
  labs(x = bquote(rho*&quot;(polygenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias in polygenic scores&quot;)+
  geom_vline(xintercept=0,color=&quot;red&quot;,linetype=&quot;dashed&quot;)

plt_bias.all</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Rare variants more appropriately correct for stratification, that much we already knew. Let’s just get the plot for “no correction”, which is what we need.</p>
<pre class="r"><code>plt1.bias = ggplot(mprs1.bias%&gt;%
                     filter(correction==&quot;pcs0&quot;),
                   aes(rlat))+
  geom_histogram()+
  theme_classic()+
  labs(x = bquote(rho*&quot;(polygenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias&quot;)</code></pre>
<p>Now plot the prediction accuracy when effect estimation and discovery in done in the same sample. We measure prediction accuracy as the correlation between polygenic scores and true genetic values.</p>
<pre class="r"><code>#calculate mean prediction accuracy across replicates
mprs1.bias.mean = mprs1.bias%&gt;%
  group_by(correction)%&gt;%
  summarize(rlat = mean(rlat),
            r2 = mean(r2))

plt1.r2 = ggplot(mprs1.bias,
                 aes(r2))+
  geom_histogram(bins=10)+
  theme_classic()+
  geom_vline(data=mprs1.bias.mean,
             aes(xintercept = r2),
             color=&quot;red&quot;,
             linetype=&quot;dashed&quot;)+
facet_wrap( ~ correction,
              labeller = as_labeller(c(
                pcs0 = &quot;No correction&quot;,
                cm = &quot;Common-PCA&quot;,
                re = &quot;Rare-PCA&quot;,
                cmre = &quot;Common + rare&quot;
              )))+
  labs(x = bquote(rho^2*&quot;(polygnenic score, genetic value)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)


plt1.r2</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="effects-re-estimated-in-a-different-sample-than-the-discovery-set" class="section level3">
<h3>Effects re-estimated in a different sample than the discovery set</h3>
<p>Now, let’s plot both bias and prediction accuracy if we ascertain variants in one sample and re-estimate in another.</p>
<pre class="r"><code>prs2 = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/revisions/prs_prediction/prs/a1_r2_p3.smooth.pcs0.all.sscore&quot;))

colnames(prs2) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;pcs0&quot;,&quot;cm&quot;,&quot;re&quot;,&quot;cmre&quot;)

mprs2 = reshape2::melt(prs2,id.vars = c(&quot;rep&quot;,&quot;IID&quot;),
                       value.name = &quot;prs&quot;,
                       variable.name = &quot;correction&quot;)

#load genetic values for individuals in the sample we are predicting - to calculate prediction accuracy
#same as the gvalue1

#add latitude info
mprs2 = merge(mprs2, pop1, by=&quot;IID&quot;)
#add genetic value
mprs2 = merge(mprs2, gvalue1, by=c(&quot;rep&quot;,&quot;IID&quot;))

#center the prs and subtract out genetic value
mprs2 = mprs2%&gt;%
  group_by(rep,correction)%&gt;%
  mutate(prs.adjusted = prs - mean(prs),
         prs.adjusted = prs.adjusted - gvalue)

#calculate the bias and prediction accuracy 
mprs2.bias = mprs2 %&gt;%
  group_by(rep,correction)%&gt;%
  summarize(rlat = cor(prs,latitude),
            r2 = cor(prs,gvalue)^2)

plt_bias.all = ggplot(mprs2.bias,aes(rlat))+
  geom_histogram(bins=10)+
  facet_wrap( ~ correction,
              labeller = as_labeller(c(
                pcs0 = &quot;No correction&quot;,
                cm = &quot;Common-PCA&quot;,
                re = &quot;Rare-PCA&quot;,
                cmre = &quot;Common + rare&quot;
              )))+
  theme_classic()+
  labs(x = bquote(rho*&quot;(polygenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias in polygenic scores&quot;)+
  geom_vline(xintercept=0,color=&quot;red&quot;,linetype=&quot;dashed&quot;)

plt_bias.all</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The bias is much smaller when the effects are re-estimated. Now plot the prediction accuracy.</p>
<pre class="r"><code>plt2.bias = ggplot(mprs2.bias%&gt;%
                     filter(correction==&quot;pcs0&quot;),
                   aes(rlat))+
  geom_histogram()+
  theme_classic()+
  labs(x = bquote(rho*&quot;(polygenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias&quot;)

#calculate mean prediction accuracy across replicates
mprs2.bias.mean = mprs2.bias%&gt;%
  group_by(correction)%&gt;%
  summarize(rlat = mean(rlat),
            r2 = mean(r2))

plt2.r2 = ggplot(mprs2.bias%&gt;%
                   filter(correction==&quot;pcs0&quot;),
                 aes(r2))+
  geom_histogram(bins=10)+
  geom_vline(xintercept=0,color=&quot;red&quot;,linetype=&quot;dashed&quot;)+
  theme_classic()+
  labs(x = bquote(rho^2*&quot;(polygnenic score, genetic value)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)

plt2.r2</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The prediction accuracy is much higher when effects are re-estimated in a different sample.</p>
</div>
<div id="effect-sizes-estimated-in-the-same-sample-as-the-discovery-set-but-with-2x-the-sample-size" class="section level3">
<h3>Effect sizes estimated in the same sample as the discovery set but with 2x the sample size</h3>
<p>Now let’s look at these plots when we discover and estimate effects in the sample but with a size twice the original sample. The question being: is the increase in prediction accuracy when we re-estimate an effect of re-estimation or due to an increase in sample size.</p>
<pre class="r"><code>prs.combined = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/revisions/combined_sample/prs/gridt100_prs_smooth.combined.all.nc.sscore&quot;))

colnames(prs.combined) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;prs&quot;)

prs.combined = merge(prs.combined,
                     gvalue1,
                     by=c(&quot;rep&quot;,&quot;IID&quot;))

prs.combined = merge(prs.combined,pop1,by=&quot;IID&quot;)
prs.combined$ascertainment = &quot;2x sample size&quot;

#center the prs and subtract out genetic value
prs.combined = prs.combined%&gt;%
  group_by(rep)%&gt;%
  mutate(prs.adjusted = prs - mean(prs),
         prs.adjusted = prs.adjusted - gvalue)

#calculate the bias and prediction accuracy 
prs.combined.bias = prs.combined %&gt;%
  group_by(rep)%&gt;%
  summarize(rlat = cor(prs,latitude),
            r2 = cor(prs,gvalue)^2)

prs.combined.bias$ascertainment = &quot;2x sample size&quot;



plt_bias.combined = ggplot(prs.combined.bias,
                           aes(rlat))+
  geom_histogram()+
  theme_classic()+
  labs(x = bquote(rho*&quot;(polygenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias in polygenic scores&quot;)

plt_bias.combined</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plt_r2.combined = ggplot(prs.combined.bias,
                           aes(r2))+
  geom_histogram()+
  theme_classic()+
  labs(x = bquote(rho*&quot;(polygenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)

plt_r2.combined</code></pre>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-7-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Combine the relevant ‘bias’ and ‘prediction accuracy’ plots into one for an easier comparison.</p>
<pre class="r"><code>mprs1.bias = mprs1.bias%&gt;%
  filter(correction==&quot;pcs0&quot;)%&gt;%
  select(rep,rlat,r2)%&gt;%
  mutate(ascertainment = &quot;discovery&quot;)%&gt;%
  ungroup()

mprs2.bias = mprs2.bias%&gt;%
  filter(correction==&quot;pcs0&quot;)%&gt;%
  select(rep,rlat,r2)%&gt;%
  mutate(ascertainment = &quot;re-estimated&quot;)%&gt;%
  ungroup()

mprs.bias = rbind(mprs1.bias, mprs2.bias,prs.combined.bias)

mprs.bias.mean = mprs.bias %&gt;%
  group_by(ascertainment)%&gt;%
  summarize(rlat = mean(rlat),
            r2 = mean(r2))

plt.bias = plt.r2 = ggplot(mprs.bias,
                 aes(rlat))+
  geom_histogram(bins=10)+
  theme_classic()+
  geom_vline(data=mprs.bias.mean,
             aes(xintercept = r2),
             color=&quot;red&quot;,
             linetype=&quot;dashed&quot;)+
facet_grid(ascertainment~.)+
  labs(x = bquote(rho^2*&quot;(polygnenic score, latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias&quot;)

plt.r2 = ggplot(mprs.bias,
                 aes(r2))+
  geom_histogram(bins=10)+
  theme_classic()+
  geom_vline(data=mprs.bias.mean,
             aes(xintercept = r2),
             color=&quot;red&quot;,
             linetype=&quot;dashed&quot;)+
facet_grid(ascertainment~.)+
  labs(x = bquote(rho^2*&quot;(polygnenic score, genetic value)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)

plt.bias + plt.r2</code></pre>
<p><img src="figure/biasvaccuracy_prsascertainment.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] patchwork_1.0.1   rprojroot_1.3-2   dplyr_1.0.2       data.table_1.13.2
[5] ggplot2_3.3.2     workflowr_1.6.2  

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       plyr_1.8.6       pillar_1.4.6     compiler_4.0.3  
 [5] later_1.1.0.1    git2r_0.27.1     tools_4.0.3      digest_0.6.27   
 [9] evaluate_0.14    lifecycle_0.2.0  tibble_3.0.4     gtable_0.3.0    
[13] pkgconfig_2.0.3  rlang_0.4.8      rstudioapi_0.11  yaml_2.2.1      
[17] xfun_0.19        withr_2.3.0      stringr_1.4.0    knitr_1.30      
[21] generics_0.1.0   fs_1.5.0         vctrs_0.3.4      grid_4.0.3      
[25] tidyselect_1.1.0 glue_1.4.2       R6_2.5.0         rmarkdown_2.5   
[29] farver_2.0.3     reshape2_1.4.4   purrr_0.3.4      magrittr_1.5    
[33] backports_1.1.10 scales_1.1.1     promises_1.1.1   ellipsis_0.3.1  
[37] htmltools_0.5.0  colorspace_1.4-1 httpuv_1.5.4     labeling_0.4.2  
[41] stringi_1.5.3    munsell_0.5.0    crayon_1.3.4    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
