<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Does fine-mapping improve prediction accuracy and reduce stratification in polygenic scores compared with clumping and thresholding?</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Demographic history mediates the effect of stratification on polygenic scores</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Arslan-Zaidi/popstructure">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Does fine-mapping improve prediction accuracy and reduce stratification in polygenic scores compared with clumping and thresholding?</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-12-22
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>popstruct_scripts/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201202code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201202)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201202code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201202)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomArslanZaidipopstructuretree99813ec398a9cad0bb1775140d09a7ab2cc431fctargetblank99813eca"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/Arslan-Zaidi/popstructure/tree/99813ec398a9cad0bb1775140d09a7ab2cc431fc" target="_blank">99813ec</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomArslanZaidipopstructuretree99813ec398a9cad0bb1775140d09a7ab2cc431fctargetblank99813eca" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/Arslan-Zaidi/popstructure/tree/99813ec398a9cad0bb1775140d09a7ab2cc431fc" target="_blank">99813ec</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rproj.user/
    Ignored:    Archived/
    Ignored:    analysis/.DS_Store
    Ignored:    burden_msprime/.DS_Store
    Ignored:    burden_msprime/generate_burden/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/burden_msprime/
    Ignored:    data/burden_msprime2/
    Ignored:    data/gwas/
    Ignored:    data/ukmap/
    Ignored:    gwas/.DS_Store
    Ignored:    gwas/grid/.DS_Store
    Ignored:    gwas/grid/tau-9/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/generate_genotypes/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/gwas/.DS_Store
    Ignored:    gwas/grid/tau100/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/generate_genotypes/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/gwas/.DS_Store
    Ignored:    gwas/ukb/.DS_Store
    Ignored:    gwas/ukb/scripts/.DS_Store
    Ignored:    optimize_migration_rate/.DS_Store
    Ignored:    optimize_migration_rate/complex_dem/.DS_Store
    Ignored:    optimize_migration_rate/grid/.DS_Store
    Ignored:    output/plots/
    Ignored:    pca_plots/.DS_Store
    Ignored:    prs/.DS_Store
    Ignored:    prs/grid/.DS_Store
    Ignored:    qqplots/.DS_Store
    Ignored:    qqplots/fixed_effects/.DS_Store
    Ignored:    qqplots/lmms/.DS_Store
    Ignored:    revisions/.DS_Store
    Ignored:    shared_scripts/gwas/
    Ignored:    shared_scripts/prs/
    Ignored:    shared_scripts/simphenotype/
    Ignored:    sib_analysis/.DS_Store
    Ignored:    sib_analysis/complex_dem/.DS_Store
    Ignored:    sib_analysis/grid/.DS_Store
    Ignored:    sib_analysis/grid/tau100/mates/
    Ignored:    simulating_genotypes/.DS_Store
    Ignored:    simulating_genotypes/grid/.DS_Store
    Ignored:    simulating_phenotypes/.DS_Store

Untracked files:
    Untracked:  analysis/biasvaccuracy_prsascertainment.Rmd
    Untracked:  analysis/biasvaccuracy_prsascertainment.nb.html
    Untracked:  analysis/plotting_prs_sib_effects.Rmd
    Untracked:  analysis/plotting_prs_sib_effects.nb.html
    Untracked:  analysis/plottingprs_distribution_gridt.Rmd
    Untracked:  analysis/plottingprs_distribution_gridt.nb.html
    Untracked:  analysis/plt_burden_association.Rmd
    Untracked:  analysis/plt_ukb_unrelated_prs.Rmd
    Untracked:  analysis/plt_ukb_unrelated_prs.nb.html
    Untracked:  analysis/prs_wt_finemapping.Rmd
    Untracked:  analysis/prs_wt_finemapping.nb.html

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd
    Modified:   burden_msprime/plt_burden_association_t100.Rmd
    Modified:   burden_msprime/plt_burden_association_t100.nb.html
    Deleted:    revisions/figuring_out_prediction_accuracy.Rmd
    Deleted:    revisions/figuring_out_prediction_accuracy.nb.html
    Deleted:    revisions/figuring_out_prediction_accuracy2.Rmd
    Deleted:    revisions/figuring_out_prediction_accuracy2.nb.html

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>One of the reviewers was interested in seeing if fine-mapping (with Susie) performed better than our clumping and thresholding in terms of bias and prediction accuracy in the polygenic score. We test this idea here.</p>
<p>Briefly, I ran susie’s algorithm with for every 100kb window individually to identify the variant with the highest posterior inclusion probability. We set a value for L=1 because we know  that there’s only one causal variant in each window. We used the original effect size (as opposed to that re-estimated by susie) to construct the polygenic scores.</p>
</div>
<div id="implementation" class="section level2">
<h2>Implementation</h2>
<p>Let’s load necessary libraries.</p>
<pre class="r"><code>library(ggplot2)
library(data.table)
library(dplyr)</code></pre>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:data.table&#39;:

    between, first, last</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>library(rprojroot)
library(patchwork)

F = is_rstudio_project$make_fix_file()

options(dplyr.summarise.inform=FALSE) </code></pre>
<p>Read the polygenic scores constructed using effect sizes of variants called by susie using summary statistics from GWAS with no correction and using rare-PCA to correct for population structure. Also, load the (true) genetic values for each individual and their longitude and latitude information.</p>
<pre class="r"><code>#no correction
prs.pcs0 = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/test/susie_prs/prs_susie_smooth.all.pcs0.sscore&quot;))

#rare-PCA
prs.re = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/test/susie_prs/prs_susie_smooth.all.re.sscore&quot;))

colnames(prs.pcs0) = colnames(prs.re)=c(&quot;rep&quot;,&quot;IID&quot;,&quot;lead&quot;,&quot;marginal&quot;)

#remove betas marginal over the windows
#we estimated these but they don&#39;t really work because it requires knowledge of the phase of the variants etc. which we have but complicated to do.
#I&#39;ve observed in many cases that the effects are very similar
prs.pcs0 = prs.pcs0 %&gt;% select(rep,IID,lead)
prs.re = prs.re %&gt;% select(rep,IID,lead)

#load genetic values
gvalues = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/test/gvalue/genos_grid_d36_m0.05_s500_t100.rmdup.test.all.gvalue.sscore.gz&quot;))
colnames(gvalues) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;dosage&quot;,&quot;gvalue&quot;)

#load pop file (lat and long information)
pop = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/iid_test.txt&quot;))


prs.pcs0$correction=&quot;pcs0&quot;
prs.re$correction=&quot;re&quot;
prs.susie= rbind(prs.pcs0,prs.re)

prs.susie = merge(prs.susie, pop)
prs.susie = merge(prs.susie, gvalues, by = c(&quot;IID&quot;,&quot;rep&quot;))


prs.susie.summary = prs.susie%&gt;%
  group_by(correction,rep)%&gt;%
  summarize(rlat = cor(lead,latitude),
            r2 = cor(lead,gvalue)^2)

prs.susie.summary = prs.susie %&gt;%
  mutate(lead.adjusted = lead - gvalue) %&gt;%
  group_by(correction,rep) %&gt;%
  summarize(rlat = cor(lead.adjusted,latitude),
            r2 = cor(lead,gvalue)^2)

prs.susie.summary.mean = prs.susie.summary %&gt;%
  group_by(correction) %&gt;%
  summarize(mean.rlat = mean(rlat),
            mean.r2 = mean(r2))</code></pre>
<p>Now let’s add the plot of when lead SNPs are used (our original strategy).</p>
<pre class="r"><code>prs.lead = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/test/prs/gridt100_prs_smooth.all.nc.sscore.gz&quot;))

colnames(prs.lead) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;dosage&quot;,&quot;pcs0&quot;,&quot;cm&quot;,&quot;re&quot;,&quot;cmre&quot;)

prs.lead = merge(prs.lead, gvalues, by = c(&quot;rep&quot;,&quot;IID&quot;))
prs.lead = merge(prs.lead, pop, by =c(&quot;IID&quot;))

prs.lead = prs.lead %&gt;% select(rep,IID,pcs0,re,gvalue,latitude)


mprs.lead = reshape2::melt(prs.lead, id.vars = c(&quot;rep&quot;,&quot;IID&quot;,&quot;gvalue&quot;,&quot;latitude&quot;),
                           value.name = &quot;prs&quot;,
                           variable.name = &quot;correction&quot;)

mprs.lead.summary = mprs.lead %&gt;%
  mutate(prs.adjusted = prs - gvalue) %&gt;%
  group_by(rep,correction) %&gt;%
  summarise(rlat = cor(prs.adjusted,latitude),
            r2 = cor(prs,gvalue)^2)

#caculate mean for each distribution
mprs.summary.lead.mean = mprs.lead.summary %&gt;%
  group_by(correction)%&gt;%
  summarize(mean.rlat = mean(rlat),
            mean.r2 = mean(r2))

mprs.lead.summary = mprs.lead.summary %&gt;% 
  select(rep,correction,rlat,r2)</code></pre>
<p>For comparison, let’s also load PRS calculated with known causal variants.</p>
<pre class="r"><code>prs.causal.p = fread(F(&quot;data/gwas/grid/genotypes/tau100/ss500/test/prs/gridt100_prs_smooth.all.c.p.sscore.gz&quot;))

colnames(prs.causal.p) = c(&quot;rep&quot;,&quot;IID&quot;,&quot;dosage&quot;,&quot;pcs0&quot;,&quot;cm&quot;,&quot;re&quot;,&quot;cmre&quot;)

prs.causal.p = merge(prs.causal.p, gvalues, by = c(&quot;rep&quot;,&quot;IID&quot;))
prs.causal.p = merge(prs.causal.p, pop, by =c(&quot;IID&quot;))

prs.causal.p = prs.causal.p %&gt;% select(rep,IID,pcs0,re,gvalue,latitude)


mprs.causal.p = reshape2::melt(prs.causal.p, id.vars = c(&quot;rep&quot;,&quot;IID&quot;,&quot;gvalue&quot;,&quot;latitude&quot;),
                           value.name = &quot;prs&quot;,
                           variable.name = &quot;correction&quot;)

mprs.causal.p.summary = mprs.causal.p %&gt;%
  mutate(prs.adjusted = prs - gvalue) %&gt;%
  group_by(rep,correction) %&gt;%
  summarise(rlat = cor(prs.adjusted,latitude),
            r2 = cor(prs,gvalue)^2)

#caculate mean for each distribution
mprs.summary.causal.p.mean = mprs.causal.p.summary %&gt;%
  group_by(correction)%&gt;%
  summarize(mean.rlat = mean(rlat),
            mean.r2 = mean(r2))

mprs.causal.p.summary = mprs.causal.p.summary %&gt;% 
  select(rep,correction,rlat,r2)</code></pre>
<p>Concatenate altogether and make plots. For this, we need to get all x axes to be on the same scale.</p>
<pre class="r"><code>prs.summary = rbind(prs.susie.summary, mprs.lead.summary, mprs.causal.p.summary)

#figure out the max range of axes so that all plots are on the same scale
xrange.rlat = range(prs.summary$rlat)

xrange.r2 = range(prs.summary$r2)</code></pre>
<p>Plot each component and combine into single plot</p>
<pre class="r"><code>plt.susie.bias = ggplot(prs.susie.summary)+
  geom_histogram(aes(rlat),bins=10)+
  facet_grid(correction~., labeller = as_labeller(
               c(pcs0 = &quot;No \n correction&quot;,
                 re = &quot;Rare-PCA&quot;)
               ))+
  theme_classic()+
  labs(x = bquote(rho~&quot;(residual PS,latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias&quot;)+
  geom_vline(xintercept = 0, color=&quot;blue&quot;, linetype = &quot;dashed&quot;)+
  geom_vline(data = prs.susie.summary.mean, 
             aes(xintercept = mean.rlat),
             linetype = &quot;dashed&quot;,
             color = &quot;red&quot;)+
  coord_cartesian(xlim = xrange.rlat)


plt.susie.r2 = ggplot(prs.susie.summary)+
  geom_histogram(aes(r2),bins=10)+
  facet_grid(correction~.,
             labeller = as_labeller(
               c(pcs0 = &quot;No \n correction&quot;,
                 re = &quot;Rare-PCA&quot;)
               ))+
  theme_classic()+
  labs(x = bquote(rho^2~&quot;(PS, genetic value)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)+
  geom_vline(data = prs.susie.summary.mean, 
             aes(xintercept = mean.r2),
             color = &quot;red&quot;,
             linetype = &quot;dashed&quot;)+
  coord_cartesian(xlim = xrange.r2)


plt.susie = plt.susie.bias + plt.susie.r2


plt.lead.bias = ggplot(mprs.lead.summary)+
  geom_histogram(aes(rlat),bins=10)+
  facet_grid(correction~., labeller = as_labeller(
               c(pcs0 = &quot;No \n correction&quot;,
                 re = &quot;Rare-PCA&quot;)
               ))+
  theme_classic()+
  labs(x = bquote(rho~&quot;(residual PS,latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias&quot;)+
  coord_cartesian(xlim = xrange.rlat)+
  geom_vline(xintercept = 0, color=&quot;blue&quot;, linetype = &quot;dashed&quot;)+
  geom_vline(data = mprs.summary.lead.mean, 
             aes(xintercept = mean.rlat),
             linetype = &quot;dashed&quot;,
             color = &quot;red&quot;)

plt.lead.r2 = ggplot(mprs.lead.summary)+
  geom_histogram(aes(r2),bins=10)+
  facet_grid(correction ~.,
             labeller = as_labeller(
               c(pcs0 = &quot;No \n correction&quot;,
                 re = &quot;Rare-PCA&quot;)
               ))+
  theme_classic()+
  labs(x = bquote(rho^2~&quot;(PS, genetic value)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)+
  coord_cartesian(xlim = xrange.r2)+
  geom_vline(data = mprs.summary.lead.mean, 
             aes(xintercept = mean.r2),
             linetype = &quot;dashed&quot;,
             color = &quot;red&quot;)


plt.lead = plt.lead.bias + plt.lead.r2


plt.causal.p.bias = ggplot(mprs.causal.p.summary)+
  geom_histogram(aes(rlat),bins=10)+
  facet_grid(correction~., labeller = as_labeller(
               c(pcs0 = &quot;No \n correction&quot;,
                 re = &quot;Rare-PCA&quot;)
               ))+
  theme_classic()+
  labs(x = bquote(rho~&quot;(residual PS,latitude)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Bias&quot;)+
  coord_cartesian(xlim = xrange.rlat)+
  geom_vline(xintercept = 0, color=&quot;blue&quot;, linetype = &quot;dashed&quot;)+
  geom_vline(data = mprs.summary.causal.p.mean, 
             aes(xintercept = mean.rlat),
             linetype = &quot;dashed&quot;,
             color = &quot;red&quot;)

plt.causal.p.r2 = ggplot(mprs.causal.p.summary)+
  geom_histogram(aes(r2),bins=10)+
  facet_grid(correction ~.,
             labeller = as_labeller(
               c(pcs0 = &quot;No \n correction&quot;,
                 re = &quot;Rare-PCA&quot;)
               ))+
  theme_classic()+
  labs(x = bquote(rho^2~&quot;(PS, genetic value)&quot;),
       y = &quot;Count&quot;,
       title = &quot;Prediction accuracy&quot;)+
  coord_cartesian(xlim = xrange.r2)+
  geom_vline(data = mprs.summary.causal.p.mean, 
             aes(xintercept = mean.r2),
             linetype = &quot;dashed&quot;,
             color = &quot;red&quot;)


plt.causal.p = plt.causal.p.bias + plt.causal.p.r2


plt.all = plt.causal.p / plt.lead / plt.susie 

plt.all</code></pre>
<p><img src="figure/prs_wt_finemapping.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] patchwork_1.0.1   rprojroot_1.3-2   dplyr_1.0.2       data.table_1.13.2
[5] ggplot2_3.3.2     workflowr_1.6.2  

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5        plyr_1.8.6        pillar_1.4.6      compiler_4.0.3   
 [5] later_1.1.0.1     git2r_0.27.1      R.methodsS3_1.8.1 R.utils_2.10.1   
 [9] tools_4.0.3       digest_0.6.27     evaluate_0.14     lifecycle_0.2.0  
[13] tibble_3.0.4      gtable_0.3.0      pkgconfig_2.0.3   rlang_0.4.8      
[17] rstudioapi_0.11   yaml_2.2.1        xfun_0.19         withr_2.3.0      
[21] stringr_1.4.0     knitr_1.30        generics_0.1.0    fs_1.5.0         
[25] vctrs_0.3.4       grid_4.0.3        tidyselect_1.1.0  glue_1.4.2       
[29] R6_2.5.0          rmarkdown_2.5     farver_2.0.3      reshape2_1.4.4   
[33] purrr_0.3.4       magrittr_1.5      backports_1.1.10  scales_1.1.1     
[37] promises_1.1.1    ellipsis_0.3.1    htmltools_0.5.0   colorspace_1.4-1 
[41] httpuv_1.5.4      labeling_0.4.2    stringi_1.5.3     munsell_0.5.0    
[45] crayon_1.3.4      R.oo_1.24.0      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
