<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Simulating heritable phenotypes</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Demographic history mediates the effect of stratification on polygenic scores</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Arslan-Zaidi/popstructure">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Simulating heritable phenotypes</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-12-22
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>popstruct_scripts/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201202code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201202)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201202code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201202)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomArslanZaidipopstructuretree99813ec398a9cad0bb1775140d09a7ab2cc431fctargetblank99813eca"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/Arslan-Zaidi/popstructure/tree/99813ec398a9cad0bb1775140d09a7ab2cc431fc" target="_blank">99813ec</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomArslanZaidipopstructuretree99813ec398a9cad0bb1775140d09a7ab2cc431fctargetblank99813eca" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/Arslan-Zaidi/popstructure/tree/99813ec398a9cad0bb1775140d09a7ab2cc431fc" target="_blank">99813ec</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rproj.user/
    Ignored:    Archived/
    Ignored:    analysis/.DS_Store
    Ignored:    burden_msprime/.DS_Store
    Ignored:    burden_msprime/generate_burden/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/burden_msprime/
    Ignored:    data/burden_msprime2/
    Ignored:    data/gwas/
    Ignored:    data/ukmap/
    Ignored:    gwas/.DS_Store
    Ignored:    gwas/grid/.DS_Store
    Ignored:    gwas/grid/tau-9/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/generate_genotypes/.DS_Store
    Ignored:    gwas/grid/tau-9/scripts/gwas/.DS_Store
    Ignored:    gwas/grid/tau100/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/generate_genotypes/.DS_Store
    Ignored:    gwas/grid/tau100/scripts/gwas/.DS_Store
    Ignored:    gwas/ukb/.DS_Store
    Ignored:    gwas/ukb/scripts/.DS_Store
    Ignored:    optimize_migration_rate/.DS_Store
    Ignored:    optimize_migration_rate/complex_dem/.DS_Store
    Ignored:    optimize_migration_rate/grid/.DS_Store
    Ignored:    output/plots/
    Ignored:    pca_plots/.DS_Store
    Ignored:    prs/.DS_Store
    Ignored:    prs/grid/.DS_Store
    Ignored:    qqplots/.DS_Store
    Ignored:    qqplots/fixed_effects/.DS_Store
    Ignored:    qqplots/lmms/.DS_Store
    Ignored:    revisions/.DS_Store
    Ignored:    shared_scripts/gwas/
    Ignored:    shared_scripts/prs/
    Ignored:    shared_scripts/simphenotype/
    Ignored:    sib_analysis/.DS_Store
    Ignored:    sib_analysis/complex_dem/.DS_Store
    Ignored:    sib_analysis/grid/.DS_Store
    Ignored:    sib_analysis/grid/tau100/mates/
    Ignored:    simulating_genotypes/.DS_Store
    Ignored:    simulating_genotypes/grid/.DS_Store
    Ignored:    simulating_phenotypes/.DS_Store

Untracked files:
    Untracked:  analysis/biasvaccuracy_prsascertainment.Rmd
    Untracked:  analysis/biasvaccuracy_prsascertainment.nb.html
    Untracked:  analysis/plotting_prs_sib_effects.Rmd
    Untracked:  analysis/plotting_prs_sib_effects.nb.html
    Untracked:  analysis/plottingprs_distribution_gridt.Rmd
    Untracked:  analysis/plottingprs_distribution_gridt.nb.html
    Untracked:  analysis/plt_burden_association.Rmd
    Untracked:  analysis/plt_ukb_unrelated_prs.Rmd
    Untracked:  analysis/plt_ukb_unrelated_prs.nb.html
    Untracked:  analysis/prs_wt_finemapping.Rmd
    Untracked:  analysis/prs_wt_finemapping.nb.html

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd
    Modified:   burden_msprime/plt_burden_association_t100.Rmd
    Modified:   burden_msprime/plt_burden_association_t100.nb.html
    Deleted:    revisions/figuring_out_prediction_accuracy.Rmd
    Deleted:    revisions/figuring_out_prediction_accuracy.nb.html
    Deleted:    revisions/figuring_out_prediction_accuracy2.Rmd
    Deleted:    revisions/figuring_out_prediction_accuracy2.nb.html

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Simulating_heritable_phenotypes.Rmd</code>) and HTML (<code>docs/Simulating_heritable_phenotypes.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/Arslan-Zaidi/popstructure/blob/99813ec398a9cad0bb1775140d09a7ab2cc431fc/analysis/Simulating_heritable_phenotypes.Rmd" target="_blank">99813ec</a>
</td>
<td>
Arslan-Zaidi
</td>
<td>
2020-12-03
</td>
<td>
organizing workflowr project
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Here, I am going to demonstrate how we simulated heritable phenotypes with environmental stratification. For illustration, I am going to use the genotypes simulated under the `recent’ structure model.</p>
<pre class="r"><code>suppressWarnings(suppressMessages({
  library(data.table)
  library(dplyr)
  library(tidyr)
  library(rprojroot)
  library(ggplot2)
}))

F = is_rstudio_project$make_fix_file()</code></pre>
<p>Read a file containing all variants and their minor allele frequency. This is important because the effects we assign to causal variants will depend on their frequency with rare variants having larger effects and common variants having smaller effects.</p>
<pre class="r"><code>set.seed(123)

#frequency file path
freq_file=&quot;data/gwas/grid/genotypes/tau100/ss500/train/genotypes/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.train.snps.frq.afreq&quot;

# load variant frequency file
p = fread(F(freq_file))

colnames(p)=c(&quot;CHROM&quot;,&quot;ID&quot;,&quot;REF&quot;,&quot;ALT&quot;,&quot;ALT_FREQS&quot;,&quot;COUNT&quot;)

p=p[,c(&quot;CHROM&quot;,&quot;ID&quot;,&quot;ALT_FREQS&quot;)]

p[, c(&quot;chr&quot;, &quot;position&quot;,&quot;ref&quot;,&quot;alt&quot;) := tstrsplit(ID, &quot;_&quot;, fixed=TRUE)]

p = p[,c(&quot;CHROM&quot;,&quot;ID&quot;,&quot;position&quot;,&quot;ALT_FREQS&quot;)]

p$position = as.numeric(p$position)</code></pre>
<p>Now sample 2,000 variants at random such that they are uniformly spaced across the genome with ~ 100Kb between them. We do this for each chromosome separately.</p>
<pre class="r"><code>#write function to do this.
sample.variant=function(df1){
  #sample first variant
  position1 = as.numeric( sample_n( df1[ position &lt; 1e5, &#39;position&#39; ], 1 ))
  #select all other variants to be at least 100kb apart
  #minimum positions for each window
  positions = position1 + seq(0,99)*1e5
  #pick variants that are further than these
  positions.adj = lapply( positions, function(x){
    ix = min( df1[position &gt; x, which =TRUE ] )
    return(df1[ix])
  })
  #return datatable
  positions.adj = bind_rows(positions.adj)
  return(positions.adj)
}

#carry this out grouped by chromosome
causal.variants &lt;- p[, sample.variant(.SD), by=CHROM]

#let&#39;s remove NAs if there are any.
causal.variants = causal.variants%&gt;%drop_na(ID)</code></pre>
<p>Now, simulate effect sizes for these variants. I’m going to model the genetic architecture after that of height (see Schoech et al. 2019) such that the effect size of a variant i,</p>
<p><span class="math display">\[\beta_i \sim N( 0, \sigma_{l}^2 . [p_i.(1-p_i)]^\alpha )\]</span></p>
<p>where:</p>
<p><span class="math inline">\(\sigma_l\)</span> = the frequency-independent component of the genetic variance associated with variant <span class="math inline">\(i\)</span>,</p>
<p><span class="math inline">\(p_i\)</span> = the frequency of the <span class="math inline">\(i_{th}\)</span> allele,</p>
<p><span class="math inline">\(\alpha\)</span> = the scaling factor which determines how the effect size is related to allele frequency.</p>
<p>Under this model, the total contribution of variant <span class="math inline">\(i\)</span> to the genetic variance is</p>
<p><span class="math display">\[\sigma_{i}^2 = \sigma_l^2.[2.p_i.(1-p_i)]^{\alpha+1}\]</span></p>
<div id="explanation" class="section level3">
<h3>Explanation</h3>
<p><span class="math display">\[\sigma_{g}^2 = Var(\beta_iX_i) = \mathbb{E}[\beta_i^2X_i^2] - \mathbb{E}[\beta_iX_i]^2\]</span> <span class="math display">\[ = \mathbb{E}[\beta_i^2].\mathbb{E}[X_i^2] - (\mathbb{E}[\beta_i].\mathbb{E}[X_i])^2\]</span> <span class="math display">\[ = \sigma_l^2.2p_i(1-p_i)[p_i(1-p_i)]^\alpha\]</span> <span class="math display">\[ = \sigma_l^2.[2p_i(1-p_i)]^{1+\alpha}\]</span></p>
<p>And the total additive genetic variance across <span class="math inline">\(M\)</span> variants is</p>
<p><span class="math display">\[\sigma_{g}^2 = \sum_{i=1}^M\sigma_{i}^2= \sum_{i=1}^M\sigma^2_{l}.[2.p_i.(1-p_i)]^{\alpha+1}\]</span></p>
<p>We would like to choose effect sizes such that the total heritability is 0.8. To do this, we must first compute the required value of <span class="math inline">\(\sigma^2_{l}\)</span>, the frequency-indpendent component of genetic variance. We use the frequencies of the variants we chose, <span class="math inline">\(\alpha=-0.4\)</span>, and the desired and <span class="math inline">\(\sigma_g^2\)</span> of 0.8 to do this.</p>
<p><span class="math display">\[\sigma^2_{l} = \frac{0.8}{ \sum_{i=1}^M[2.p_i.(1-p_i)]^{1-0.4} }\]</span></p>
<pre class="r"><code>#calculate the independent component of variance required
sigma2_l = 0.8 / sum( sapply( causal.variants$ALT_FREQS,function(x){
  beta= ( 2*x*(1-x)) ^ (1-0.4)
  return(beta)
}))</code></pre>
<p>Next, we sample the effect sizes for each variant.</p>
<pre class="r"><code>#sample maf-dependent effects using the model above
causal.variants$beta = sapply( causal.variants$ALT_FREQS , function(x){
  beta = rnorm( 1 , mean = 0, sd = sqrt(sigma2_l * (2*x*(1-x))^-0.4 ))
})

#let&#39;s calculate sigma2_g to confirm that the total genetic variance is indeed 0.8
sigma2_g = sum( mapply(function(b,p){ b^2* 2*p*(1-p) }, causal.variants$beta, causal.variants$ALT_FREQS))

print(paste(&quot;sigma2_g : &quot;,round(sigma2_g,3)))</code></pre>
<pre><code>[1] &quot;sigma2_g :  0.76&quot;</code></pre>
<p>Now, we use these effects and the genotypes to calculate the simulated genetic value for each <span class="math inline">\(j^{th}\)</span> individual in PLINK (computation not shown):</p>
<p><span class="math inline">\(g_j = \sum_{i=1}^M \beta_iX_{ij}\)</span></p>
<p>Read the genetic values and the `pop’ file, a file which contains which deme an individual belongs to and their latitude and longitude values.</p>
<pre class="r"><code>gvalue_file = &quot;data/gwas/grid/genotypes/tau100/ss500/train/genotypes/genos_gridt100_l1e7_ss750_m0.05_chr1_20.rmdup.train.1.thinned_100kb.gvalue.sscore&quot;

popfile = &quot;data/gwas/grid/genotypes/tau100/ss500/iid_train.txt&quot;

prs=fread(F(gvalue_file))
colnames(prs)&lt;-c(&quot;IID&quot;,&quot;dosage&quot;,&quot;gvalue&quot;)

sample_size=nrow(prs)

#load file containing the deme id and latitude and longitude for each individual
pop=fread(F(popfile))

#add this info to prs file
prs=merge(prs, pop, by=&quot;IID&quot;, sort=FALSE)</code></pre>
<p>Add environmental component to the genetic values such that: a) the phenotype is either smoothly (North-South gradient) or sharply distributed, and b) the phenotype has a heritability of 0.8.</p>
<pre class="r"><code>#smooth effect - latitude (OG effect)
prs$smooth = sapply(prs$latitude,
                      function(x){
                        rnorm(n = 1,
                              mean = (x + 1)/3,
                              sd = sqrt(1 - 0.8))})


#sharp environmental effect
prs$sharp = sapply(prs$deme,
                 function(x){
                   if(x == 2){
                     rnorm(n = 1,
                           mean = 2,
                           sd = sqrt(1 - 0.8)) }else{
                             rnorm(n = 1,
                                   mean = 0,
                                   sd = sqrt(1 - 0.8))
                           }})</code></pre>
<p>Scale the effects such that the heritability is 0.8.</p>
<pre class="r"><code>prs$sharp = scale( prs$sharp , scale = T) * sqrt( 1 - 0.8)
prs$smooth = scale( prs$smooth, scale = T) * sqrt(1 - 0.8)

#add prs to each of the environmental effects
prs = prs %&gt;%
  mutate(smooth = gvalue + smooth,
         sharp = gvalue + sharp)

#the correlation between the phenotype and the genetic values should be ~ 0.8
print(paste(&quot;h2 (smooth) :&quot;,
            round(cor(prs$gvalue,
                      prs$smooth)^2,2)))</code></pre>
<pre><code>[1] &quot;h2 (smooth) : 0.8&quot;</code></pre>
<pre class="r"><code>print(paste(&quot;h2 (sharp) :&quot;,
            round(cor(prs$gvalue,
                      prs$sharp)^2,2)))</code></pre>
<pre><code>[1] &quot;h2 (sharp) : 0.8&quot;</code></pre>
<p>Now let’s plot the spatial distribution of the mean phenotype within each deme.</p>
<p>For the smooth effect:</p>
<pre class="r"><code>prs_mean = prs%&gt;%
  group_by(deme,longitude,latitude)%&gt;%
  summarize(smooth = mean(smooth),
            sharp = mean(sharp))</code></pre>
<pre><code>`summarise()` regrouping output by &#39;deme&#39;, &#39;longitude&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>midpoint_sm = mean(prs_mean$smooth)

prsplot_sm=ggplot(prs_mean)+
  geom_tile(aes(longitude,latitude,fill=smooth))+
  scale_fill_gradient2(high = &quot;#fc8d59&quot;,
                       mid = &quot;#ffffbf&quot;, 
                       low = &quot;#91bfdb&quot;,
                       midpoint = midpoint_sm)+
  theme_bw()+
  labs(x=&quot;Longitude&quot;,
       y=&quot;Latitude&quot;,
       fill=&quot;Phenotype\n(smooth)&quot;)

prsplot_sm</code></pre>
<p><img src="figure/Simulating_heritable_phenotypes.Rmd/plot_smooth-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>For the sharp effect:</p>
<pre class="r"><code>midpoint_shp = mean(prs_mean$sharp)

prsplot_shp=ggplot(prs_mean)+
  geom_tile(aes(longitude,latitude,fill=sharp))+
  scale_fill_gradient2(high = &quot;#fc8d59&quot;,
                       mid = &quot;#ffffbf&quot;, 
                       low = &quot;#91bfdb&quot;,
                       midpoint = midpoint_shp)+
  theme_bw()+
  labs(x=&quot;Longitude&quot;,
       y=&quot;Latitude&quot;,
       fill=&quot;Phenotype\n(sharp)&quot;)

prsplot_shp</code></pre>
<p><img src="figure/Simulating_heritable_phenotypes.Rmd/plot_sharp-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.3.2     rprojroot_1.3-2   tidyr_1.1.2       dplyr_1.0.2      
[5] data.table_1.13.2 workflowr_1.6.2  

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       pillar_1.4.6     compiler_4.0.3   later_1.1.0.1   
 [5] git2r_0.27.1     tools_4.0.3      digest_0.6.27    gtable_0.3.0    
 [9] evaluate_0.14    lifecycle_0.2.0  tibble_3.0.4     pkgconfig_2.0.3 
[13] rlang_0.4.8      rstudioapi_0.11  yaml_2.2.1       xfun_0.19       
[17] withr_2.3.0      stringr_1.4.0    knitr_1.30       generics_0.1.0  
[21] fs_1.5.0         vctrs_0.3.4      grid_4.0.3       tidyselect_1.1.0
[25] glue_1.4.2       R6_2.5.0         rmarkdown_2.5    farver_2.0.3    
[29] purrr_0.3.4      magrittr_1.5     whisker_0.4      scales_1.1.1    
[33] backports_1.1.10 promises_1.1.1   ellipsis_0.3.1   htmltools_0.5.0 
[37] colorspace_1.4-1 httpuv_1.5.4     labeling_0.4.2   stringi_1.5.3   
[41] munsell_0.5.0    crayon_1.3.4    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
